### PR TITLE
Unidling status updates

### PR DESCRIPTION
**Pull request by andyhd**
# What

Apologies for the massive diff. It's mostly the updated vendor packages.

* **Added Server Sent Events during unidling**
  This lets users see the progress of the app unidling in a dynamically updated webpage.

* **Refactor for testing**
  Testing this should be easier by making an `interface` for the Kubernetes API client and creating a mock implementation to use in the tests.

* **Refactor to recommended best practices**
  [Eg: Don't Panic!](https://github.com/golang/go/wiki/CodeReviewComments)

* **Switch to use `go mod` for dependencies**
  Packages are vendored, so it doesn't cause any problems. Best to keep up to date with the recommended way of doing things.

# How to review

1. checkout and use `make run` to run the unidler on `localhost:8080`
2. browse to http://localhost:8080/
3. see an "Unidling localhost:8080" page, with dynamically updated list of Server Sent Event messages - this should be one item "Failed unidling 'localhost:8080': can't find ingress for host 'localhost:8080'"
4. browse to http://localhost:8080/?host=IDLED_HOST where IDLED_HOST is the hostname of an idled app, eg: http://localhost:8080/?host=andy-test.apps.dev.mojanalytics.xyz
5. see an "Unidling IDLED_HOST" page, with SSE messages dynamically updating

I'm still not very experienced in Go, so some things might not be idiomatic - please comment!

### CODE REVIEW

File: templates/index.html [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+			eventItem.textContent = e.data
+			events.appendChild(eventItem);
+		};
+
+		source.onerror = function () {
+			console.log("EventSource failed.");
+		};
+	</script>
+</body>
+</html>
```
Comment (r4vi) [inline templates/index.html]: shouldn't it `window.refresh()` after a short delay once a successfully unidled message is received?

**Comment by r4vi** [comment]
also errors when running `make docker-image`:
```
Step 6/12 : RUN make static
 ---> Running in 55a0b27ee9a3
 > Checking dependencies...
go: modules disabled inside GOPATH/src by GO111MODULE=auto; see 'go help modules'
make: *** [Makefile:19: dependencies] Error 1
The command '/bin/sh -c make static' returned a non-zero code: 2
make: *** [docker-image] Error 2
/home/ravi/code/moj/analytics/analytics-platform-go-unidler/Makefile:23: recipe for target 'docker-image' failed
```

File: templates/index.html (hunk 1 above)
Comment (andyhd) [reply templates/index.html]: oh, you mean redirect to the unidled app? yeah, i will do that. i didn't flesh out the unidling page yet.

File: app.go [hunk 2]
```diff
@@ -22,138 +23,195 @@ type (
...
+		return
+	}
+	if err = a.RemoveFromUnidlerIngress(); err != nil {
+		return
+	}
+	if err = a.RemoveIdledMetadata(); err != nil {
+		return
+	}
+	a.setStatus("Ready")
+	return nil
```
[the remark this answers is not in the document]
Comment (andyhd) [reply app.go]: I have to return something, because the function signature says so.

File: k8s.go [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+	"k8s.io/client-go/kubernetes"
+	_ "k8s.io/client-go/plugin/pkg/client/auth/oidc"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+type (
+	// KubernetesWrapper defines a minimal set of methods for a Kubernetes API
+	// client for the Unidler - this allows easier testing
+	KubernetesWrapper interface {
```
[the remark this answers is not in the document]
Comment (andyhd) [reply k8s.go]: I didn't know it existed - thanks for the tip

**Comment by andyhd** [comment]
I haven't committed any test code yet, but I have tested it manually

File: templates/index.html (hunk 1 above)
Comment (r4vi) [reply templates/index.html]: exactly that but keep in mind one of the issues of the old unilder was the checking of the unidled app using XHR. That didn't work if the unidled app redirected you to auth0 because the request would be blocked by CORS

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+## static: Build static binary.
+static: dependencies
+	@echo " > Building binary..."
+	@${GO} build -a -installsuffix cgo -ldflags '-extldflags "-static"' -o ${PROJECTNAME} .
+
+## run: Run the unidler.
+run: static
+	@echo " > Starting unidler"
+	@$(PROJECTNAME)
```
Comment (r4vi) [inline Makefile]: had to change to `./@$(PROJECTNAME)` to make it run

File: Makefile [hunk 5]
```diff
@@ -0,0 +1,61 @@
+PROJECTNAME := $(shell basename "$(PWD)")
+
+GOBASE := $(shell pwd)
+#GOPATH := $(GOBASE)/vendor:$(GOBASE)
+GOBIN := $(GOBASE)/bin
+GOFILES := $(wildcard *.go)
+
+STDERR := /tmp/.$(PROJECTNAME)-stderr.txt
+PID := /tmp/.$(PROJECTNAME).pid
```
Comment (xoen) [inline Makefile]: What are these `PID` and `STDERR` variables about?

Comment (andyhd) [reply Makefile]: I copied this Makefile from another project and haven't cleaned it all up yet. I'll remove these.

**Comment by xoen** [comment]
FYI: I had a look at Go modules (to not confuse with Go packages) but when I looked the k8s libraries didn't play well with it. If I remember correctly they were discussing it in a GH issue. 

Having said that, it's great if you managed to make it work.

File: app.go [hunk 6]
```diff
@@ -2,216 +2,219 @@ package main
...
-		return err
-	}
+// Unidle performs the actions to unidle an app
+func (a *App) Unidle() (err error) {
+	a.status <- "Pending"
 
-	err = a.removeFromUnidlerIngress()
-	if err != nil {
-		return err
+	if err = a.SetReplicas(1); err != nil {
```
Comment (xoen) [inline app.go]: Having the function call and the error check in the same line as if condition is not idiomatic in Go: https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow

Comment (andyhd) [reply app.go]: 👍

File: main.go [hunk 7]
```diff
@@ -1,20 +1,94 @@
...
+	ingressClassName, ok = os.LookupEnv("INGRESS_CLASS_NAME")
+	if !ok {
+		ingressClassName = "nginx"
+	}
+	home, ok := os.LookupEnv("HOME")
+	if !ok {
+		log.Fatalf("Couldn't determine HOME directory, is $HOME set?")
+	}
+	var err error
+	k8s, err = NewKubernetesAPI(filepath.Join(home, ".kube", "config"))
```
Comment (xoen) [inline main.go]: Unless I'm missing something here, you should be able to use the shorthand declaration/assignment `:=` and avoid the `var err error` line.

File: unidler.go [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+func (u *UnidlerServer) HandleRequests() {
+	log.Print("Unidler started")
+
+	for {
+		select {
+		// if a request to unidle a host is received and that host is not
+		// currently being unidled, start unidling - it should be impossible to
+		// receive a request for an already unidled app
+		case host := <-u.requests:
+			u.startUnidling(host)
```
Comment (xoen) [inline unidler.go]: (Bear in mind I just started to have a look and still wrapping my head around as the code got more complex...)

Correct me if I'm wrong, but this would be blocking so although the `HandleRequests()` method will run in the background when a `host` string is sent to the `u.requests` channel it would starting unidling that host **syncronously** and therefore blocking the other requests.

Is that the expected behaviour? I don't know if you'd like to start the unindling in a separate go routine here and in the `stopUnidling` case.

File: main.go (hunk 7 above)
Comment (andyhd) [reply main.go]: Using the shorthand assignment here would create a shadowing `k8s` variable in the function scope. This is probably a sign that the `k8s` variable should not be global. I'll try to refactor this.

File: unidler.go [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+		namespace  string
+		requests   chan string
+		sse        SseSender
+		taskEnd    chan string
+	}
+
+	// UnidleTask represents the progress of unidling an app
+	UnidleTask struct {
+		app     *App
+		host    string
```
[the remark this answers is not in the document]
Comment (andyhd) [reply unidler.go]: Yes, but it's not great to break encapsulation. I guess that's an OO mindset though.

File: unidler.go [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+type (
+	// Unidler provides a method to unidle an app
+	Unidler interface {
+		EndTask(*UnidleTask)
+		Unidle(string)
+	}
+
+	// UnidlerServer represents the unidler server
+	UnidlerServer struct {
+		inProgress map[string]*UnidleTask
```
Comment (xoen) [inline unidler.go]: So this map is keeping the state of what's being idled and what not.
I'm not very keen on state as it was one of the reason why the Python version of the unidler was sometimes not working.

I'm just starting looking but maybe we can still run more than one instances of the go-unidler as unidling is idempotent (as the steps involved are idempotent).

Another question is, do we need to keep this state? If the unidling method use the channel to update on progress and the SeeBroker send the events to the client does it matter if we keep track of who is idling or not? (sure the user could refresh the page and trigger the unidling of the same app twice+ but that's not a problem right?)

Oh, another maybe stupid question: Can we run more of one instances of this app or only one `SeeBroker` instance can run and talk with all the clients?

File: main.go (hunk 7 above)
Comment (xoen) [reply main.go]: Good point as it looks like you only use it in the `main()` function (directly)

File: unidler.go (hunk 8 above)
Comment (andyhd) [reply unidler.go]: `startUnidling` calls `UnidleTask.Run()` in a goroutine - so won't that run concurrently? `HandleRequests` waits for a request, calls `startUnidling` which synchronously creates the task and fetches the app ingress and deployment, then calls `Run()` asynchronously. It seems to work the way I expect, but I haven't really tested concurrently unidling multiple apps.

File: unidler.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (xoen) [reply unidler.go]: There is something off about this interface but I can't really tell if it's just personal preference or something more objective.

For example, the `UnidlerServer` keeps an instance of the `SseSender` (the `SseBroker`) which makes sense to me as you would expect to have only one of those. But then the `UnidleTask` has also an `sse` attribute, why?

I'm wondering if both the methods `Unidle()`/`EndTask()` (`Cancel()`/`CancelUnidle`?) could just receive the host string, or maybe the `EndTask()` method could receive the app.

In fact now that I see the implementation of `EndTask()` it just uses the host and ignore the rest:

```
func (u *UnidlerServer) EndTask(task *UnidleTask) {
	u.taskEnd <- task.host
}
```

Maybe I'm missing something (e.g. you added the rest for testing purposes, etc...)

File: unidler.go (hunk 10 above)
Comment (andyhd) [reply unidler.go]: Yes, it is keeping state. Unless we actually are running multiple Unidler processes (and I don't see why we would want to), I don't think it is a problem. I thought about keeping state purely in the Kubernetes resources (eg: step 1 in unidling an app is to add a "Currently Unidling" label to the app ingress and deployment), but I don't see how to avoid a race condition.
Perhaps you are right that multiple concurrent attempts to unidle the same app are not a problem.
What is the need for running more than one Unidler?

File: unidler.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (andyhd) [reply unidler.go]: Good points. I'm not entirely happy with the interface either. I'm still coming at this with a OO mindset and it's tripping me up in places.

File: unidler.go (hunk 8 above)
Comment (xoen) [reply unidler.go]: Oh I see. I didn't get to its implementation but I see it spawn a go routing by doing `go task.Run()`

I was almost tempted to suggest to run the whole `startUnidling()` in a go routine but I think that may increate the chance of getting a race condition.

Bear in mind go maps are not safe for concurrent use (unless you use a mutex of some sort). I'm not 100% sure there are no race conditions here as potentially two go routines could read/write from the map at the same time 💥

File: unidler.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (xoen) [reply unidler.go]: Don't worry, don't see these comments as a critic. I think you're picking up Go great from what I see (maybe too keen on interfaces and go channels, but I know how tempting it is, they are great features when used properly)

I'm wrapping my head around these changes and writing down these comments as it goes, so probably most of these will be stupid as I don't have the whole picture yet.

File: templates/index.html [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
...h78AaMNkEFKV6rSlm5UmiTVgQEkEICaBoABBtDBSD+qUZaC9JkyyAAEFoABDOTAqBj4AwYWoIEYoLSlLE1pRqXaTB4wIAdY/81qVgWBVQyc4AY75alUM7pMD4DAqDmgQVrVSgNDrAADNIiBB6BK17omkwcSgGta94rVSBx1B2Pl6UuDGVS0qnWvap2EVzPw1KjalbAawOph2ZqDSuRgASwQrEd7KQMWGHWyWW2rZWGgA7o6lqW45EEN+ApaTGCgBZrdbCsbcIDQalW0l8DADRprWqi2UgY6WIBksSrcxGIirbHt6Ck7O9zbVjYTGGAAb0+L0eVm4LCXRexzMbECDSR3sHQMKl+1mlZNXJa61D0lD3awV+EWd7uXSGtgY7tcFpB3uOaFQW+p+9NP8kAHkdVuDlbwhwcMmBJHDcB3yfpbGRigBi2AwUplF0CDCid2BUe971FpsIOp7peunOXBSBvQAB2U2MQsYEEMGMAAEECgBi9+cQAyMOIFJ1OzJc2xiMM63d5Gs6OBpepT6ctR9BYiEAAh+QQFCgA/ACx/AEgAjADCAIUAQWPedSkASWsIYYTeODkYdZznliHeWTEIVXMQbZTeSTHnhinnoiEQaZQAUXMQaYzeQTEIXYTnfSkQYYzecSkIXXsQcZznqhgASWMATXPnniHeZTEIVXveUTHnjiHeQTkARWPneSkATWsQYYTePDnnmiHeXTEQcZTeTTHniinnpiEIUXPeRTHngikQZYwYcZznqiEIWXsAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAG/8CfcEgsGo0gkEAgEmBAx6h0Sq1ar9hsFiR6vArgxCCpLZvP6DQVFCt8we8Eh6yu2+93EMftPhX8bw8OdHiFhodIAoB/fF9+biMiUIiUlXVsfIuOjQUJMT+TlqKjVSAPnF+pjIsJK4SksLGmYJqofGAFLpKxvKQgLqi1my9vBSMCob3KhSAjcIyZnH6AxCefycvZaCArjKmawtBfDa7Y2udYv9GruKqLi7m76PNXXOEv07bvcCOv9P9IuuDK507VMG+dYvgDSE9JEgwxEjzDhzDcvgLkFjJc5kQJBmRcRuR79s4gQmJgXCDbeC4JSGRPkgj8Y/DRSXYEK6xk2cshsv+fHmMiOMWOlsmjt+QAMMeTkkeHH32CePIkRs1oF
```
Comment (xoen) [inline templates/index.html]: Do you need this variable after this point? It doesn't seem like it's used. If not I think you could just avoid the `else` block. 

If I understand correctly host is coming from the server (which renders this template) and it needs to end up in the `host` url param in `url`. Maybe something like this?

```js
  var host = urlparams.get("host");
  if (!host) {
    host = "{{.}}";
  }
  url += '?host=' + host;
```

Or something like that.

Comment (xoen) [inline templates/index.html]: Stupid questions: 
1) Where is `EventSource` is coming from?
2) Why you register `onerror` using `source.onerror` but the success handler is registered using ` source.addEventListener("success", function (e) // ...`?

Comment (andyhd) [reply templates/index.html]: It's implemented in the browser: https://developer.mozilla.org/en-US/docs/Web/API/EventSource

"success" is a custom event type. Errors are a default type.

Comment (andyhd) [reply templates/index.html]: Good point - I was using it later in the script, but replaced it with the template interpolation '{{.}}'. This code here was only for testing purposes anyway. I'll remove it.

File: unidler.go (hunk 10 above)
Comment (xoen) [reply unidler.go]: > What is the need for running more than one Unidler?
K8s works better with stateless application, we could want to have more instances running (currently there are 3 if I remember correctly) and also bear in mind that pods could be killed and re-schedule at any point (e.g. k8s scales down on the number of nodes). 

Having stateless applications also mean which upgrade k8s and drain the pods from a node to an upgraded one happens without the user noticing it, as for example one pod is re-scheduled while two are still responding to requests, etc...

(This is also one of the reasons RStudio instances are a pain, as each user has only one instance running. If for whatever reason this single pod gets unresponsive or we upgrade the helm chart or upgrade k8s the user see a `503 Service Unavailable`)

I'm still reviewing these changes, if I may have some ideas when I understand better how this SSE broker works, etc...

File: templates/index.html (hunk 11 above)
Comment (xoen) [reply templates/index.html]: Uh magical. Thanks, I thought it was some kind of 3rd party library, I didn't realise this was a standard API 👍

File: main.go [hunk 12]
```diff
@@ -1,20 +1,94 @@
...
+	fmt.Fprint(w, "Still OK")
+}
+
+func unidle(w http.ResponseWriter, req *http.Request) {
+	if req.URL.Path != "/" {
+		return
+	}
+	log.Printf("HTTP request received for %s", req.URL.Path)
+
+	tmpl, err := template.ParseFiles("templates/index.html")
```
Comment (xoen) [inline main.go]: Potentially you could move this "expensive" operation which doesn't need to happen at every request at application startup. It **may be** a good use case for [`init()` functions](https://golang.org/doc/effective_go.html#init) (I never used them but they should just run before `main()`, but you may need to declare both variables to avoid shadowing)

File: sse.go [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+		channels  map[string]*Channel
+		addClient chan *Client
+		delClient chan *Client
+	}
+)
+
+func (m *Message) String() string {
+	s := ""
+
+	if m.id != "" {
```
Comment (xoen) [inline sse.go]: Does anyone set `id` or `retry` on any instance of `Message`?

File: sse.go [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+	ch.lastMessage = msg
+	for c := range ch.clients {
+		c.send <- msg.String()
+	}
+}
+
+// NewSseBroker constructs a new SSE server and starts it running
+func NewSseBroker() *SseBroker {
+	s := &SseBroker{
+		make(map[string]*Channel),
```
[the remark this answers is not in the document]
Comment (andyhd) [reply sse.go]: 👍

File: main.go (hunk 12 above)
Comment (andyhd) [reply main.go]: 👍

File: sse.go (hunk 13 above)
Comment (andyhd) [reply sse.go]: Not at the moment, but they are part of the specification, so when I changed from just sending a string to sending a Message struct (so that I could set the event type), I figured it should probably implement the specification.

**Comment by andyhd** [comment]
I agree with you on avoiding this becoming stateful. I think we can go back to treating the unidling process as idempotent and not caring about multiple concurrent unidle requests. There's no need for redis or any shared state.

**Comment by xoen** [comment]
👆 Another thing that I thought and I forgot to mention is that we may use the status from k8s to determine when the deployment is ready (similarly to what we do in the `WaitForDeployment()`). That could help to simplify. I know the devil is in the details but I thought that may help.

File: sse.go (hunk 13 above)
Comment (xoen) [reply sse.go]: I see. The reason I was asking is it because if you can include these even if blank without breaking the client you could use a backtick string, e.g.:

```go
func (m *Message) String() string {
    return fmt.Sprintf(`id: %s
event: %s
retry: %d
data: %s

`, m.id, m.event, m.retry, m.data)
	}
```
or  something like that ([Go Playground](https://play.golang.org/p/Or59SOoh-3H))

Don't worry too much about it, I don't mind the `if`s, but I thought you wanted to know about this way of writing long string literals :)

**Comment by andyhd** [comment]
A test is still failing at the moment - just trying to work out if k8s.io fake is not working the way I expect

**Comment by xoen** [comment]
FYI: I'm getting a checksum mismatch on `k8s.io/client-go` when it's trying to install the dependencies:

```sh
$ make dependencies
 > Checking dependencies...
go: verifying k8s.io/client-go@v10.0.0+incompatible: checksum mismatch
	downloaded: h1:h3fciHPG0O5QEzATTFoRw/YGtDsU6pxrMrAhxiTtcq0=
	go.sum:     h1:F1IqCqw7oMBzDkqlcBymRq1450wD0eNqLE9jzUrIi34=
make: *** [dependencies] Error 1
```

I'm not familiar enough with go modules to understand how this could be fixed (a part from just messing with the checksum in `go.sum`).

File: sse.go [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+
+	// Channel is a named channel which can broadcast messages to a list of Clients
+	Channel struct {
+		name        string
+		clients     map[*Client]bool
+		lastMessage *Message
+	}
+
+	// SseSender provides a method to send a SSE event to a specified channel
+	SseSender interface {
```
Comment (xoen) [inline sse.go]: Assuming SSE is an acronym and the correct spelling is SSE (and it seems the case by looking at [the MDN page](https://developer.mozilla.org/en-US/docs/tag/SSE)) this should be `SSESender`.

I can't find where this is documented right now but you can see this convention all over the standard library, e.g. [`ServeHTTP`](https://golang.org/pkg/net/http/#HandlerFunc.ServeHTTP), [`http.StatusOK`](https://golang.org/pkg/net/http/#pkg-constants), [`MarshalJSON`](https://golang.org/pkg/encoding/json/#RawMessage.MarshalJSON), [`HTMLEscape`](https://golang.org/pkg/text/template/#HTMLEscape), etc, etc...

File: unidler.go [hunk 16]
```diff
@@ -0,0 +1,76 @@
+package main
+
+import (
+	"fmt"
+	"log"
+)
+
+type (
+	// UnidleTask represents the progress of unidling an app
+	UnidleTask struct {
```
Comment (xoen) [inline unidler.go]: I just realised that this file is all about `UnidlerTask`, although all these types are part of the same package (`main`) and it doesn't really matter (from the compiler perspective), we may want to rename this file to `unidler_task.go` for predictability.

File: Dockerfile [hunk 17]
```diff
@@ -1,14 +1,30 @@
...
+COPY go.mod ./
+
+ENV GO111MODULE=on
 
 # NOTE: statically compiled as final image is based on "scratch"
-RUN CGO_ENABLED=0 GOOS=linux go build -a -installsuffix cgo -ldflags '-extldflags "-static"' -o go-unidler .
+RUN make static
+
+#FROM builder AS test
+#RUN make test
```
Comment (xoen) [inline Dockerfile]: Don't know if it's feasible but we may also want to run `go vet` which can highlight some common mistakes (e.g. variables not used, etc...)

File: app.go [hunk 18]
```diff
@@ -2,216 +2,193 @@ package main
...
+	app = &App{
+		host: host,
+		k8s:  k8s,
 	}
-
-	err = a.getDeployment()
+	app.ingress, err = k8s.IngressForHost(host)
 	if err != nil {
-		return err
+		return nil, err
```
Comment (xoen) [inline app.go]: In general it's a good practice to add some context to errors before returning them. As these can be low level errors. e.g.

```go
app.ingress, err = k8s.IngressForHost(host)
if err != nil {
    return nil, fmt.Errorf("failed to get ingress for '%s': %s", host, err)
}
// ...
```

This is very useful to investigate when something goes wrong as you get the chain of events that happened.

That's [the approach across the standard library](https://www.mycodesmells.com/post/smart-golang-error-handling).

File: logger.go [hunk 19]
```diff
@@ -1,10 +0,0 @@
-package main
-
-import (
-	"log"
-	"os"
-)
-
-func NewLogger(prefix string) *log.Logger {
-	return log.New(os.Stdout, prefix, log.LstdFlags|log.LUTC|log.Lshortfile)
```
Comment (xoen) [inline logger.go]: Mmm, we lost this? I know it's a bit ugly but it's very useful to see where messages come from and the time in logs otherwise it's pretty much a bunch of random characters without context.

If you don't like to have to pass around this costom logger instance maybe we can use a global variable which I know it's not great.

Don't know, have a think, you may think it's not worth the pain. When I was looking at the logs to keep an eye on it, I personally found the time and filename useful.

File: unidler.go [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+
+func (t *UnidleTask) sendStatus(data string) {
+	t.sse.SendSse(t.host, &Message{data: data})
+}
+
+// Fail ends the task with a failure status
+func (t *UnidleTask) Fail(err error) {
+	msg := &Message{
+		event: "error",
+		data:  fmt.Sprintf("%s", err),
```
Comment (xoen) [inline unidler.go]: I suppose doesn't matter much but you could use `data: fmt.Sprint(err)` or `data: err.Error()`

(BTW, I'm adding all these silly comments as I think you may find it interesting to know some of these Go "tricks" or way of doing things - don't feel like you have to change everything)

File: app.go [hunk 21]
```diff
@@ -2,216 +2,193 @@ package main
...
+	if labelExists || annotationExists {
+		delete(a.deployment.Annotations, IdledAtAnnotation)
+		delete(a.deployment.Labels, IdledLabel)
+		_, err := a.k8s.UpdateDeployment(a.deployment)
+		if err != nil {
+			if strings.Contains(err.Error(), "object has been modified") {
+				dep, err := a.k8s.Deployment(a.ingress)
+				if err == nil {
+					a.deployment = dep
+					return a.RemoveIdledMetadata()
```
Comment (xoen) [inline app.go]: Wow, is this recursive? Is it to retry until it succeeds? We can potentially end up with an "infinite" loop here right?

File: sse.go [hunk 22]
```diff
@@ -0,0 +1,173 @@
...
+		http.Error(w, "Streaming not supported", http.StatusInternalServerError)
+		return
+	}
+
+	w.Header().Set("Content-Type", "text/event-stream")
+	w.Header().Set("Cache-Control", "no-cache")
+	w.Header().Set("Connection", "keep-alive")
+
+	// add the browser as a new client of the channel for the specified host
+	host := req.Host
```
Comment (xoen) [inline sse.go]: Isn't this the same as `getHost()`?

File: sse.go [hunk 23]
```diff
@@ -0,0 +1,173 @@
...
+func (s *SseBroker) SendSse(channel string, msg *Message) {
+	if channel == "" {
+		for _, ch := range s.channels {
+			ch.sendMessage(msg)
+		}
+	} else if _, ok := s.channels[channel]; ok {
+		s.channels[channel].sendMessage(msg)
+	} else {
+		ch := newChannel(channel)
+		s.channels[ch.name] = ch
```
Comment (xoen) [inline sse.go]: Possibly you can write `s.channels[ch.name] = newChannel(channel)`, etc...

File: sse.go [hunk 24]
```diff
@@ -0,0 +1,173 @@
...
+	for {
+		select {
+		case c := <-s.addClient:
+			ch, exists := s.channels[c.channel]
+			if !exists {
+				ch = newChannel(c.channel)
+				s.channels[ch.name] = ch
+			}
+			ch.clients[c] = true
+			if ch.lastMessage != nil {
```
Comment (xoen) [inline sse.go]: Sorry if it's a stupid question but why do we send the last message when a client connects?

File: sse.go [hunk 25]
```diff
@@ -0,0 +1,173 @@
...
+				c.send <- ch.lastMessage.String()
+			}
+
+		case c := <-s.delClient:
+			if ch, exists := s.channels[c.channel]; exists {
+				close(c.send)
```
Comment (xoen) [inline sse.go]: Mmmm, watch out for this `close` as as far as I remember if a go routine will try to send to this after you close it, it will panic. As far as I can tell don't close a channel is not a big deal as it will eventually be garbage collected (e.g. when no one reference to this address anymore).

In this case it seems like close it is necessary tho as the `ServeHTTP()` is will wait forever otherwise so we'll rely on the client to terminate the request. 

I'm not aware of any easy way to check if a channel is not closed before sending to it, but have a think to see if it could potentially cause problems in our case.

File: sse.go [hunk 26]
```diff
@@ -0,0 +1,173 @@
...
+	flusher.Flush()
+
+	for msg := range c.send {
+		fmt.Fprintf(w, msg)
+		flusher.Flush()
+	}
+}
+
+// SendSse broadcasts a message to a channel (or all channels)
+func (s *SseBroker) SendSse(channel string, msg *Message) {
```
Comment (xoen) [inline sse.go]: Did you call this `SendSse()` instead of `sendMessage()` or `send()` to avoid confusion with the other method? It is conceptually the same thing (but one send to a specific client the other to a whole channel)

File: sse.go (hunk 15 above)
Comment (andyhd) [reply sse.go]: Sure. I read a thing that said the Go way was SseSender and not SSESender, but I think that looks wrong anyway.

File: Dockerfile (hunk 17 above)
Comment (andyhd) [reply Dockerfile]: 👍

File: logger.go (hunk 19 above)
Comment (andyhd) [reply logger.go]: The standard logger does prepend the timestamp. I agree the prefix may be useful. I'll put it back.

File: app.go (hunk 18 above)
Comment (andyhd) [reply app.go]: 👍

File: unidler.go (hunk 20 above)
Comment (andyhd) [reply unidler.go]: Yes, I knew that. Will change it.

File: sse.go (hunk 24 above)
Comment (andyhd) [reply sse.go]: There are a couple of situations where the client can connect after an SSE was sent:
* The browser might request '/events/' after a delay (network lag) and miss the "Waiting for deployment" message, and then not be able to show a status for a few seconds
* The unidling process failed and the last message is an error message.
This may have only been possible in testing, but I think it doesn't hurt to keep it.

File: sse.go (hunk 23 above)
Comment (andyhd) [reply sse.go]: Assuming you mean `s.channels[channel] = newChannel(channel)`. 👍

Comment (andyhd) [reply sse.go]: But then I'd have to write `s.channels[channel].lastMessage = msg` - seems a bit verbose.

File: sse.go (hunk 15 above)
Comment (xoen) [reply sse.go]: Eheh, you could change the interface to just `Sender` :P 

```go
Sender interface {
    Send(string, *Message)
}
```

File: app_test.go [hunk 27]
```diff
@@ -0,0 +1,133 @@
...
+
+	err = app.RemoveIdledMetadata()
+	assert.Nil(t, err)
+
+	dep, _ = api.Deployment(ing)
+	_, labelExists := dep.Labels[IdledLabel]
+	_, annotationExists := dep.Annotations[IdledAtAnnotation]
+	assert.False(t, labelExists, "Idled label not removed")
+	assert.False(t, annotationExists, "Idled annotation not removed")
+	assert.Equal(t, expectedReplicas, *dep.Spec.Replicas)
```
Comment (xoen) [inline app_test.go]: Is this assertion to check that re-read the deployment works as expected? I'm asking because we check the number of replicas above as well.

Comment (andyhd) [reply app_test.go]: This is to check that the replicas are not changed again. Probably not necessary.

File: Makefile [hunk 28]
```diff
@@ -0,0 +1,49 @@
+PROJECTNAME := $(shell basename "$(PWD)")
+PORT := 8000
+GO := CGO_ENABLED=0 go
+
+
+default: static
+
+dependencies:
+	@echo " > Checking dependencies..."
+	@GO111MODULE=on $(GO) mod vendor
```
Comment (xoen) [inline Makefile]: I'm having another look at Go modules. So my understanding so far is that when you run a `go mod vendor" it looks at the `go.mod` (and maybe `go.sum`) and will download the dependencies and put in the `/vendor` directory.

We check in this directory as this is all you need to reproduce the build with the exact same versions of all the packages.

If that's the case (and we do check in the `/vendor` directory), what's the need for running this `go mod vendor`? Is it in case we didn't check in the `/vendor` directory? I'm actually now noticing that the copy of `/vendor` is currently commented out. I'm basically trying to see if I'm understanding Go modules.

Comment (xoen) [reply Makefile]: So, this is a [good article from Dave Cheney on the subject](https://dave.cheney.net/2018/07/16/using-go-modules-with-travis-ci).

If understand correctly, the `/vendor` directory was used in Go <1.11. If using Go 1.11+:

- if module is inside GOPATH Go will not use Go module behaviour
- if somewhere else it will use Go module behaviour

To force Go to use new Go module behaviour regardless from source code location set `GO111MODULE=on`

I think `go mod vendor` is mainly necessary to vendor the dependencies in case someone wants to use the module with Go <1.11.

Bear in mind that in the `Dockerfile` we do put the source code inside the `$GOPATH` so set `GO111MODULE=on` is a good thing to be on the safe side and force Go to use `go.mod`.

Basically, if I understand correctly I think `go mod vendor` is unnecessary and slow, but I may be wrong.

Comment (andyhd) [reply Makefile]: I think you are correct, we do not need to run `go mod vendor`.

We can either
1. Store the vendored modules in the repository as we are currently doing, and then copy the `vendor` directory into the Docker container, in which case the Makefile does not need to run `go mod vendor`. This would make for a faster image build.
OR
2. We can set `GO111MODULE=on` in the Dockerfile, which will download the modules when building the image. This would allow us to remove the modules from the repository (usually a good idea).

Comment (xoen) [reply Makefile]: I see you went for option 1 (vendored dependencies), it's probably what I would have done as it helps with reproducibility.
